website/blog/2025-03-19-20-must-try-mongodb-queries-operations-ferretdb.md

```diff
@@ -0,0 +1,436 @@
+Also, `_id` is included by default, so we explicitly set it to `0` in the example so as to remove it.
+This makes our queries faster and responses smaller – perfect when dealing with large datasets.
+
+## Aggregation operations
+
+### 16. Aggregation pipeline stages
+
+Aggregation pipelines let you process and transform data in stage where each stage refines the result.
+This is essential for analytics, reporting, and summarizing large datasets.
+
+Let's say you need to find how many books belong to the "Classic" genre; `$match` filters books that have "Classic" in their `genres` array and `$count` gives the total number of matching documents.
+
+```js
+db.books.aggregate([
+  { $match: { genres: 'Classic' } },
+  { $count: 'classic_books' }
+])
+```
+
+Find out more about aggregation pipelines in FerretDB [here](https://docs.ferretdb.io/usage/aggregation/).
+
+### 17. Run analytical operations on FerretDB with `$group` and `$avg`
+
+Say you want to analyze the average book rating per genre, which is a common use case for dashboards, trend analysis, or user recommendations.
+
+```js
+db.books.aggregate([
+  {
+    $group: {
+      _id: '$genres',
+      average_rating: { $avg: '$analytics.average_rating' }
+    }
+  }
+])
+```
+
+## Geospatial queries
+
+### 18. Find books published in a specific location
+
+Some books aren't just about different places – they're from them.
+Geospatial queries help you query location-based data, like books published in a specific city or country.
+If a publisher's location is stored as GeoJSON points, FerretDB lets you query by geography instead of just text.
+
+What if you want to search for books published in a specific city.
+Instead of manually stating their locations, you can store and query precise geographic coordinates using `$geoWithin`:
+
+Using London's longitude and latitude (`[-0.1276, 51.5072]`), let's run some queries to find books published within a `1km` radius.
+Note that the distance is in radians, so we need to convert it to the Earth's radius.
+Since the Earth's radius is about 6378.1km, we divide `1km` by the Earth's radius to get the distance in radians.
+
+```js
+db.books.find({
+  'publication.publisher.location.geolocation': {
+    $geoWithin: {
+      $centerSphere: [[-0.1276, 51.5072], 1 / 6378.1]
+    }
+  }
+})
+```
+
+## 19 Monitor operations with `currentOp`
+
+Sometimes, your database operations become slow, and you wonder what's causing the delay.
+`currentOp()` lets you inspect active operations, find long-running queries, and even terminate problematic commands before they overwhelm your instance.
+
+To actually see something in `currentOp()`, we need a long-running operation.
+Let's simulate a background insert that adds a document every 5 seconds.
+
+```js
+while (true) {
+  db.books.insertOne({
+    title: 'Live Insert Test',
+    price: { value: 9.99, currency: 'USD' }
+  })
+  sleep(5000) // Wait 5 seconds before inserting again
+}
```

Comment:
I don't know how this could work, it's not a long running operation. The operation itself like you say runs every 5 seconds, and how can `currentOp` command capture something that? I suggest removing this example.

Reply:
Yeah, sleep pauses the `mongosh` itself, not FerretDB